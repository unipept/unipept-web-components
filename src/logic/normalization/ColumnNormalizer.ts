import { Normalizer } from "./Normalizer";

/**
 * Normalises a grid on a column-per-column basis. The largest and smallest value per column are selected, after which
 * all other values in this column are normalised with respect to these values.
 *
 * @author Pieter Verschaffelt
 */
export default class ColumnNormalizer implements Normalizer {
    public normalize(data: number[][]): number[][] {
        if (data.length === 0 || data[0].length === 0) {
            return data;
        }

        const output: number[][] = [];

        for (const _ of data) {
            output.push([]);
        }

        for (let col = 0; col < data[0].length; col++) {
            // Find the minimum and maximum value by iterating over every value in the current column
            const minMax: number[] = Array.from(
                Array(data.length).keys()
            )
                .map(
                    (row) => data[row][col]).reduce((acc, current) => [Math.min(acc[0], current),
                    Math.max(acc[1], current)], [Infinity, -Infinity],
                );
            const min = minMax[0];
            const max = minMax[1];

            for (let row = 0; row < data.length; row++) {
                if (max - min !== 0) {
                    output[row].push((data[row][col] - min) / (max - min));
                } else {
                    output[row].push(0);
                }
            }
        }

        return output;
    }

    public toString(): string {
        return "ColumnNormalizer";
    }
}
