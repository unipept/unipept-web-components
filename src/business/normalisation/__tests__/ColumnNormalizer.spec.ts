import ColumnNormalizer from "@/business/normalisation/ColumnNormalizer";

describe("ColumnNormalizer", () => {
    it("normalizes a complete grid correctly", () => {
        const data = [
            [4, 4, 2, 1],
            [1, 2, 1, 0],
            [4, 3, 0, 4]
        ];

        runTestForGrid(data);
    });

    it("normalizes a grid with negative values correctly", () => {
        const data = [
            [4, 4, -2, 1],
            [-1, 2, 1, 0],
            [-4, 3, 0, 4]
        ];

        runTestForGrid(data);

    });

    it("does not crash on all-zero grids", () => {
        const normalizer = new ColumnNormalizer();
        const data = [
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0]
        ];

        const normalizedResult = normalizer.normalize(data);
        // There should no longer be negative values in the result, since everything is mapped onto the [0, 1] interval.
        expect(normalizedResult).toMatchSnapshot();
    });
});

function runTestForGrid(data: number[][]) {
    const normalizer = new ColumnNormalizer();

    const normalizedResult: number[][] = normalizer.normalize(data);
    expect(normalizedResult).toMatchSnapshot();

    // Check if each column does indeed contain a 1 and 0 value (since the normalisation is performed on a column-b
    // basis).
    let allContainMaximum = true;
    let allContainMinimum = true;
    for (let i = 0; i < data[0].length; i++) {
        let max = false;
        let min = false;
        for (let j = 0; j < data.length; j++) {
            if (data[j][i] === 1) {
                max = true;
            }
            if (data[j][i] === 0) {
                min = true;
            }
        }
    }

    expect(allContainMaximum).toBeTruthy();
    expect(allContainMinimum).toBeTruthy();
}
