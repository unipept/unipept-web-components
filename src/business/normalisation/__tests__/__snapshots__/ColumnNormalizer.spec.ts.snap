// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnNormalizer does not crash on all-zero grids 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`ColumnNormalizer normalizes a complete grid correctly 1`] = `
Array [
  Array [
    1,
    1,
    1,
    0.25,
  ],
  Array [
    0,
    0,
    0.5,
    0,
  ],
  Array [
    1,
    0.5,
    0,
    1,
  ],
]
`;

exports[`ColumnNormalizer normalizes a grid with negative values correctly 1`] = `
Array [
  Array [
    1,
    1,
    0,
    0.25,
  ],
  Array [
    0.375,
    0,
    1,
    0,
  ],
  Array [
    0,
    0.5,
    0.6666666666666666,
    1,
  ],
]
`;
